cassovary-core/src/main/scala/com/twitter/cassovary/graph/Traverser.scala

```diff
@@ -119,166 +124,420 @@ class RandomTraverser(graph: Graph, dir: GraphDir, homeNodeIds: Seq[Int],
+{
+  val onlyOnce = false
+
+  override lazy val limits = walkLimits
+
+  override lazy val prevNbrCounter = apwPrevNbrCounter
+
+  override def nodePriority = GraphTraverserNodePriority.FIFO
+
+  override def shouldBeEnqueued(color: NodeColor.Color): Boolean = true
+}
+
+/**
+ * Traverses in DFS order (every node can be visited only once).
+ * @param graph the graph to traverse on
+ * @param dir direction in which to traverse
+ * @param homeNodeIds the ids of nodes that the walk starts from
+ * @param walkLimits limiting parameters
+ */
+final class DepthFirstTraverser(val graph: Graph, val dir: GraphDir, val homeNodeIds: Seq[Int],
+                                walkLimits: WalkLimits = WalkLimits())
+    extends QueueBasedTraverser
+    with PathLengthTracker
+    with DepthTracker
+    with BoundedIterator[Node]
+{
+  private val log = Logger.get()
+
+  override lazy val limits = walkLimits
+
+  override def nodePriority = GraphTraverserNodePriority.LIFO
+
+  override def visitNode(node: Node): Unit = {
+    log.ifTrace { "visiting %d, nbrCount=%d, depth=%d, distance=%d".format(node.id,
+      node.neighborCount(dir), depth(node.id).get, distance(node.id).get) }
+    super.visitNode(node)
+  }
+
+  /**
+   * Seeks the next node that should be visited. We may need to skip
+   * some nodes, because it is possible that we add a node to the queue twice.
+   *
+   * Takes already visited nodes from the queue.
+   * Does not take the next node from the queue (in order to allow hasNext checking).
+   * */
+   @tailrec
+   override def findNextNodeToVisit(): Option[Int] = {
+    if (queue.isEmpty) {
+      None
     } else {
-      val childId = nd.neighborIds(dir)(off)
-      if (onlyOnce && seenBefore(childId)) {
-        // try the next child
-        firstNotYetVisitedNode(nd, off + 1)
+      val next = queue.firstInt()
+      val visitedBefore = coloring.get(next) match {
+        case NodeColor.Visited => true
+        case _ => false
+      }
+      if (visitedBefore) {
+        queue.dequeueInt()
+        findNextNodeToVisit()
       } else {
-        storeForVisit(nd.id, off)
-        Some(NodeDesc(childId, distance(nd.id).get + 1))
+        Some(next)
       }
     }
   }
 
-  def next = {
-    var NodeDesc(id, off) = stack.remove(stack.size - 1)
-    assert(off == -1)
-    val topnd = getExistingNodeById(graph, id)
-    val nextUnvisitedNode = firstNotYetVisitedNode(topnd, off + 1)
-    nextUnvisitedNode match {
-      case Some(NodeDesc(nodeId, dist)) => visitPotentialNode(nodeId, -1, dist)
-      case None => // do nothing
+  // can't limit number of nodes added to the queue, because we skip some nodes
+  override def limitAddedToQueue(nodes: Seq[Int]): Seq[Int] = nodes
+
+  override def shouldBeEnqueued(color: NodeColor.Color): Boolean = color != NodeColor.Visited
+}
+
+
+/**
+ * A queue based traverser trait that keeps track of visiting distance from `homeNodeIds`.
+ */
+trait PathLengthTracker extends QueueBasedTraverser {
```

Comment:
So let's leave PathLengthTracker and FinishingTimeTracker as optional when mixed in. Details coming soon.